### ISSUE TITLE
Incorrect report of unbound variable

### ISSUE THREAD

**Issue by svenzorgdoc** [microsoft/pyright#7822]
**Describe the bug**
In a loop with a fixed number of iterations, pyright does not understand that the loop will run that many times, causing things that happen in the loop to sometimes not happen in pyright's view.

**Code or Screenshots**

```python
def test() -> list[int]:
    for i in (0, 1, 2):
        result = [i]
    return result
```

**VS Code extension or command-line**
```shell
$ npx pyright@latest test.py
Need to install the following packages:
  pyright@1.1.361
Ok to proceed? (y) 
/tmp/test.py
  /tmp/test.py:4:12 - error: "result" is possibly unbound (reportPossiblyUnboundVariable)
1 error, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Pyright is working as intended here, and I don't consider this a bug. In general, static type checkers do not know how many times a `for` loop will execute. Pyright builds its code flow graph assuming that all `for` loops will execute zero or more times and performs analysis accordingly.

You should avoid referencing variables that are assigned only within a loop outside of that loop. The recommended approach here is to initialize the variable prior to the loop with some initial value.

If you don't like pyright's behavior here, you can disable the `reportPossiblyUnboundVariable` check. It was separated from the `reportUnboundVariable` check specifically so you can disable the "possibly unbound" case without disabling the "definitely unbound" case.

**Comment by wmanley** [comment]
Previously: #844.  Discussion: https://github.com/microsoft/pyright/discussions/2033
